icepyx/core/auth.py

```diff
@@ -0,0 +1,152 @@
+import copy
+import datetime
+import warnings
+
+import earthaccess
+
+class AuthenticationError(Exception):
+    '''
+    Raised when an error is encountered while authenticating Earthdata credentials
+    '''
+    pass
+
+
+class EarthdataAuthMixin():
+    """
+    This mixin class stores properties and methods related to logging into Earthdata. 
+    It can be inherited by any other class that requires authentication.
+    
+    The class provides 3 properties: auth, session, and s3login_credentials. The method
+    earthdata_login() is included for backwards compatibility.
+    
+    The class can be created without any initialization parameters, and the properties will
+    be populated when they are called. It can alternately be initialized with an 
+    earthaccess.auth.Auth object, which will then be used to create a session of 
+    s3login_credentials as they are called.
+    
+    Parameters
+    ----------
+    auth : earthaccess.auth.Auth, default None
+        Optional parameter to pass initialize an object with existing credentials
+    
+    Examples
+    --------
+    >>> a = EarthdataAuthMixin()
+    >>> a.session # doctest: +SKIP
+    >>> a.s3login_credentials # doctest: +SKIP
+    """
+    def __init__(self, auth=None):
+        self._auth = copy.deepcopy(auth)
+        # initializatin of session and s3 creds is not allowed because those are generated
+        # from the auth object
+        self._session = None
+        self._s3login_credentials = None
+        self._s3_initial_ts = None  # timer for 1h expiration on s3 credentials
```

Comment:
Trying to think through the arguments for and against having these as private variables. Are they public in earthaccess? If so, it might be good to follow that practice here.

Reply:
In the `__init__` method these variables need to be private, as this is the machinery for setting them as a read-only property for the user. We internally set the private versions (ex. `_session`) and then make them publicly available by using the `@property` decorator. The public accessors (ex. `.session`) are behind the scenes accessing the private one -- notice that each of the return statements for the properties returns the value of the private variable. As developers we change the value of the private versions, but the user only accesses the public versions.